### ISSUE TITLE
Alternative implementation/correction of HGT example

### ISSUE THREAD

**Issue by fmellomascarenhas** [dmlc/dgl#2346]
## 🚀 Feature
DGL has a HGT (Heterogeneous Graph Transformer) example created by @acbull and @VoVAllen: https://github.com/dmlc/dgl/tree/ac282a5e35ca7f8abdc8c14bde0d7db1305e2fe9/examples/pytorch/hgt

If you check the model.py file on line 69, you will see that softmax is done per edge type, instead of considering all edges, **due to DGLs framework limitation**, and this seems to be a bad thing (explained below). I wonder if there is a way to implement the same model, but doing softmax across all of the edges of the node at the same time.

## Motivation

![image](https://user-images.githubusercontent.com/23004953/99103606-b232e700-2594-11eb-971f-844ab48773f1.png)

Lets suppose that one destination node received messages from 3 types of nodes: Orange, Yellow and Blue. Now lets suppose that Blue nodes are noise and they shouldn't be added to the final message. If softmax were applied to all edges, then all blue nodes could have an attention set to 0. However, since the softmax across all blue nodes have to sum to 1, then blue nodes will always be summed to the destination node.

**Comment by jermainewang** [comment]
Hi, you are right that the current implementation applies softmax relation wise. DGL does support softmax across all of the neighborhood by converting a heterograph to a homogeneous one. Below is a demo code:
```python
import dgl
import dgl.nn as dglnn
from dgl.ops import edge_softmax
import torch

def softmax_among_all(hetgraph, weight_name):
    """Normalize edge weights by softmax across all neighbors.

    Parameters
    -------------
    hetgraph : DGLGraph
        The input heterogeneous graph.
    weight_name : str
        The name of the unnormalized edge weights.
    """
    # Convert to homogeneous graph; DGL will copy the specified data to the new graph.
    g = dgl.to_homogeneous(hetgraph, edata=[weight_name])
    # Call DGL's edge softmax
    g.edata[weight_name] = edge_softmax(g, g.edata[weight_name])
    # Convert it back; DGL again copies the data back to a heterogeneous storage.
    hetg2 = dgl.to_heterogeneous(g, hetgraph.ntypes, hetgraph.etypes)
    # Assign the normalized weights to the original graph
    for etype in hetg2.canonical_etypes:
        hetgraph.edges[etype].data[weight_name] = hetg2.edges[etype].data[weight_name]
```

You could try it out with the following example and print out the normalized weight:
```python
# create an example heterograph
hg = dgl.heterograph({
    ('user', 'plays', 'game'): ([0, 1, 0], [1, 2, 0]),
    ('store', 'sells', 'game'): ([0, 1, 1, 1], [2, 0, 1, 2])
})
# init edge weight
hg.edges['plays'].data['w'] = torch.ones((3,))
hg.edges['sells'].data['w'] = torch.ones((4,))

softmax_among_all(hg, 'w')

# Print out the result
print(hg.edges['plays'].data['w'])
print(hg.edges['sells'].data['w'])
```

**Comment by fmellomascarenhas** [comment]
@jermainewang, thank you so much for your solution.

I ran a test and it worked. I added the _softmax_among_all_ before _G.multi_update_all_, and removed the _softmax_ from _reduce_fun_.

However, **time per epoch doubled**, probably because the graph is recreated twice. I didn't see any major problem with memory though. Based on your experience, do you think that the speed problem can be addressed somehow?

**Comment by jermainewang** [comment]
Since softmax is removed from the reduce func, one way to speed it up is to replace it with builtin reduce function. You could check the usage of `u_mul_e` in the GAT example.
